src/main/kotlin/io/snyk/plugin/snykcode/SnykCodeBulkFileListener.kt

```diff
@@ -28,50 +29,71 @@ class SnykCodeBulkFileListener : SnykBulkFileListener() {
-                            TextDocumentIdentifier(virtualFile.toLanguageServerURL()),
-                            virtualFile.readText()
-                        )
-                    languageServer.textDocumentService.didSave(param)
-                }
 
-                VirtualFileManager.getInstance().asyncRefresh()
-                runInEdt { DaemonCodeAnalyzer.getInstance(project).restart() }
-            }
-        })
+        runAsync {
+            if (ApplicationManager.getApplication().isDisposed) return@runAsync
+            if (project.isDisposed) return@runAsync
+            if (DumbService.getInstance(project).isDumb) return@runAsync
 
+            val languageServerWrapper = LanguageServerWrapper.getInstance()
+            if (languageServerWrapper.isDisposed() || !languageServerWrapper.isInitialized) return@runAsync
+
+            val languageServer = languageServerWrapper.languageServer
+            val cache = getSnykCachedResults(project)?.currentSnykCodeResultsLS
+            val filesAffected = toSnykFileSet(project, virtualFilesAffected)
+            val index = ProjectFileIndex.getInstance(project)
+
+            for (file in filesAffected) {
+                val virtualFile = file.virtualFile
+                if (!shouldProcess(virtualFile, index, project)) continue
+                cache?.remove(file)
+                val param =
+                    DidSaveTextDocumentParams(
+                        TextDocumentIdentifier(virtualFile.toLanguageServerURL()),
+                        virtualFile.readText()
+                    )
+                languageServer.textDocumentService.didSave(param)
+            }
+            VirtualFileManager.getInstance().asyncRefresh()
+            DaemonCodeAnalyzer.getInstance(project).restart()
+        }
     }
 
-    private fun shouldProcess(file: VirtualFile): Boolean {
-        val inCache = debounceFileCache.getIfPresent(file.path)
+    private fun shouldProcess(file: VirtualFile, index: ProjectFileIndex, project: Project): Boolean {
+        var shouldProcess = false
+        val application = ApplicationManager.getApplication()
+        if (application.isDisposed) return false
+        if (project.isDisposed) return false
+        if (DumbService.getInstance(project).isDumb) shouldProcess = false
 
-        return if (inCache != null) {
-            logger<SnykCodeBulkFileListener>().info("not forwarding file event to ls, debouncing")
-            false
-        } else {
-            debounceFileCache.put(file.path, true)
-            logger<SnykCodeBulkFileListener>().info("forwarding file event to ls, not debouncing")
-            true
+        application.runReadAction {
+            val inCache = debounceFileCache.getIfPresent(file.path)
+            if (inCache != null) {
+                shouldProcess = false
+            } else {
+                debounceFileCache.put(file.path, true)
+                if (index.isInContent(file) && !isInBlacklistedParentDir(file)) {
```

Comment:
IntelliJ sends file updates for ALL open projects and their files, so `isInContent` is used to filter for the current project. I now wonder, if this is correct.